lib/HLSDownloader.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * Created by AsherS on 8/20/15.
+ */
+
+/**
+ * Created by AsherS on 8/11/15.
+ */
+
+var Q = require('q');
+var _ = require('underscore');
+var path = require('path');
+var url = require('url');
+var downloadUtils = require('./DownloadUtils');
+var fileUtils = require('./FileUtils');
+var hlsUtils = require('./HLSUtils');
+
+var HLSDownloader = (function () {
+
+    function HLSDownloader(m3u8Url, destPath) {
+        this.streamUrl = m3u8Url;
+        this.destPath = destPath;
+    }
+
+    var downloadIndefinitely = function (url, dest) {
+        downloadPlaylist(url, dest)
+            .then(function(targetDuration){
+                console.log('results: ' + targetDuration);
+            })
+            .catch(function(err){
+                console.log('error ' + err);
+            })
+            .done();
+
+        setTimeout(function () {
+            console.log('-------------------');
+            downloadIndefinitely(url, dest);
+        }, 10000);  //todo, configuration?
+    };
+
+    var downloadPlaylist = function (m3u8Url, dest) {
+
+        var parsedUrl = url.parse(m3u8Url);
+        var baseUrl = parsedUrl.href.substring(0, parsedUrl.href.lastIndexOf('/'));
+        var playlistName = path.basename(parsedUrl.pathname);
+        var playlistDest = path.join(dest, playlistName);
+
+        var downloadedM3u8 = undefined; //TODO initialize
+        var targetDuration = 0;
+
+        return downloadUtils.downloadFile(m3u8Url, playlistDest)
+
+            .then(function finishedDownload() {
+                return hlsUtils.readM3u8File(playlistDest);
+            })
+            .then(function(m3u) {
+                targetDuration = m3u.get("targetDuration");
+                downloadedM3u8 = m3u;
+                return fileUtils.listDirFiles(dest);
+            })
+            .then(function(files) {
+                //get all files that are in playlist but not in dir
+                //TODO - null pointer on playlistItem
+                return _.chain(downloadedM3u8.items.PlaylistItem).map(function(i){
```

Comment:
i should address it here only if the chunks in the m3u8 are urls and not just chunklist names. i haven't seen any such case so far. what do u think?